From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ian Wahbe <ian@wahbe.com>
Date: Tue, 24 Oct 2023 15:25:29 -0700
Subject: [PATCH] Avoid panic in getRawConfig


diff --git a/fastly/base_fastly_service.go b/fastly/base_fastly_service.go
index dd5bd859..8424d051 100644
--- a/fastly/base_fastly_service.go
+++ b/fastly/base_fastly_service.go
@@ -188,7 +188,11 @@ func validateUniqueNames(block string) func(_ context.Context, rd *schema.Resour
 	return func(_ context.Context, rd *schema.ResourceDiff, _ any) error {
 		names := make(map[string]int)
 
+		// Avoid panic in rd.GetRawConfig()
 		c := rd.GetRawConfig()
+		if c.IsNull() {
+			return nil
+		}
 		m := c.AsValueMap()
 		s, ok := m[block]
 		if ok {
